menpo3d/morphablemodel/fitter.py

```diff
@@ -0,0 +1,382 @@
+import sys
+import time
+import numpy as np
+import menpo.io as mio
+import menpo3d.io as m3dio
+import matplotlib.pyplot as plt
+from menpo.feature import gradient
+from menpo.image import Image
+from menpo3d.rasterize import GLRasterizer
+from menpo.transform import Homogeneous, UniformScale
+from .lmalign import retrieve_view_projection_transforms
+from .detectandfit import detect_and_fit
+from .derivatives import compute_pers_projection_derivatives_shape_parameters, \
+                              compute_ortho_projection_derivatives_shape_parameters, \
+                              compute_ortho_projection_derivatives_warp_parameters, \
+                              compute_pers_projection_derivatives_warp_parameters, \
+                              compute_texture_derivatives_texture_parameters, \
+                              compute_projection_derivatives_warp_parameters, \
+                              compute_projection_derivatives_shape_parameters
+
+
+class MMFitter(object):
+    r"""
+    Class for defining a 3DMM fitter.
+
+    """
+    def __init__(self, mm):
+        self._model = mm
+        self.result = None
+        self.rasterized_result = None
+        self.rasterizer = None
+        self.errors = []
+
+    @property
+    def mm(self):
+        r"""
+        The 3DMM model.
+
+        """
+        return self._model
+
+    @property
+    def get_result(self):
+        r"""
+        The 3DMM model fitting result.
+
+        """
+        return self.result
+
+    # TODO
+    def fit(self, image_path, n_alphas=100, n_betas=100, n_tris=1000, camera_update=False, max_iters=100):
+
+        # PRE-COMPUTATIONS
+        # ----------------
+        # Constants
+        threshold = 1e-3
+        # Projection type: 1 is for perspective and 0 for orthographic
+        projection_type = 1
+
+        # Import the image 
+        image = mio.import_image(image_path)
+        
+        # Do face detection and landmark fitting
+        detect_and_fit(image)
```

Comment:
we may wish to use ground truth landmarks that are provided to us on the input image rather than re-running the detect and fit which may fail.

the pattern we use here to is provide a `group=None` key to the method in question. The group is the name of the landmark group that we expect to have landmarks placed at. A good example of the pattern is [`Image.crop_to_landmarks()`](http://docs.menpo.org/en/stable/api/menpo/image/Image.html#menpo.image.Image.crop_to_landmarks).

Notice how in the [implementation](https://github.com/menpo/menpo/blob/master/menpo/image/base.py#L1226) the group key is used to access `.landmarks[group]`.

if group is `None`, and there is one and only one landmark group, the one landmark group is returned.

Else, the group needs to a string to specify the landmarks we are referring too.

Now the calling of the `.fit` method (including the change to take an image not a path, see above) would be:

```py
image = mio.import_image(path)
detect_and_fit(image)
# because we know that detect and fit sets this group name for the landmarks it finds
fitter.fit(image, group='ibug68', ...) 
```

If say, the user wants to use the provided lenna image (with provided [landmarks](https://github.com/menpo/menpo/blob/master/menpo/data/lenna.ljson)) they would do:
```py
image = mio.import_builtin_asset.lenna_png()
# no need to detect and fit as we have ground truth 2D landmarks!
# but we do need to tell the fitter the name of the landmarks to use...
fitter.fit(image, group='LJSON', ...) 
# actually, there is only one landmark group, so it's unambiguous, so you could provide None...
fitter.fit(image, group=None, ...) 
# ...and as this is the default, you don't need to provide it at all
fitter.fit(image, ...) 
# all three options are the same!
```